File: python/camera_ui_sdk/sensor/tamper.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class TamperProperty(StrEnum):
    """Property names of a tamper sensor."""

    Detected = "detected"
    """Whether tampering is detected."""


class TamperSensorProperties(TypedDict):
    """Property values of a tamper sensor."""

    detected: bool


class TamperPropertyChangeData(TypedDict):
    """Property change payload emitted on TamperSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a TamperProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class TamperSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a tamper sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Tamper

    @property
    def onPropertyChanged(self) -> Observable[TamperPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[TamperProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class TamperSensor(Sensor[TamperSensorProperties, TStorage, str], Generic[TStorage]):
    """Tamper sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Tamper Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({TamperProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Tamper

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report tampering detection state.

        Args:
            value: True when tamper is currently detected.

        Example:
            ```python
            tamper.setDetected(True)
            ```
        """
        self._write_state({TamperProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""

File: node/src/sensor/tamper.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a tamper sensor.
 *
 * @internal
 */
export enum TamperProperty {
  /** Whether tampering is detected. */
  Detected = 'detected',
}

/**
 * Property values of a tamper sensor.
 *
 * @internal
 */
export interface TamperSensorProperties {
  [TamperProperty.Detected]: boolean;
}

/** Read-only proxy interface for a tamper sensor. */
export interface TamperSensorLike extends SensorLike {
  readonly type: SensorType.Tamper;
  readonly onPropertyChanged: Observable<PropertyChangeOf<TamperSensorProperties>>;

  getValue(property: TamperProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Tamper sensor. */
export class TamperSensor<TStorage extends object = Record<string, any>> extends Sensor<TamperSensorProperties, TStorage> {
  readonly type = SensorType.Tamper;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Tamper Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [TamperProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report tampering detection state.
   *
   * @param value - True when tamper is currently detected.
   *
   * @example
   * ```ts
   * tamper.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [TamperProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link TamperSensor}. */
export const tamperMeta = defineSensor({
  type: SensorType.Tamper,
  category: SensorCategory.Sensor,
  assignmentKey: 'tamper',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [TamperProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: TamperProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [TamperProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: TamperProperty.Detected,
    commandProperty: TamperProperty.Detected,
    deviceClass: 'tamper',
  },
});
